src/action/ActionManager.js

```diff
@@ -0,0 +1,301 @@
+import ActionMap from "./ActionMap.js";
+
+import Filter from "../filter/Filter.js";
+import ReverseActiveFilter from "../filter/ReverseActiveFilter.js";
+import ReverseAxisFilter from "../filter/ReverseAxisFilter.js";
+
+import Device from "../hardware/Device.js";
+import Gamepad from "../hardware/Gamepad.js";
+import Keyboard from "../hardware/Keyboard.js";
+
+import InputSource from "../input/InputSource.js";
+import MouseInputSource from "../input/MouseInputSource.js";
+import GamepadInputSource from "../input/GamepadInputSource.js";
+import KeyboardInputSource from "../input/KeyboardInputSource.js";
+
+/**
+ * If new ActionManager.constructor's loadDefaults param is true then these action maps are loaded
+ */
+const defaultActionMaps = [
+  { name: "default-flat", url: "/js/action-input/default/playing-flat-action-map.json" }
+  /*
+  { name: "default-portal", url: "../default/playing-portal-action-map.json" },
+  { name: "default-immersive", url: "../default/playing-immersive-action-map.json" }
+  */
+];
+
+/**
+ *
+ * TODO figure out how to create links/aliases from specific hw inputs to generic inputs
+ */
+export default class ActionManager {
+  constructor(loadDefaults = true) {
+    this.queryInputPath = this.queryInputPath.bind(this);
+    /**
+     * input semantic path (with wildcards) -> action listener {function}
+     * @type {Map<string, function>}
+     */
+    this._actionListeners = new Map();
+
+    /**
+     * semantic path -> {@link InputSource}
+     * @type {Map<string, InputSource>}
+     */
+    this._inputSources = new Map();
+
+    /**
+     * full semantic path -> {@link Filter}
+     * @type {Map<string, Filter>}
+     */
+    this._filters = new Map();
+
+    /**
+     * name -> {@link ActionMap}
+     * @type {Map<string, ActionMap>}
+     */
+    this._actionMaps = new Map();
+
+    /**
+     * zero or more {@link ActionMap}s may be active at any time
+     * @type {Map<string, ActionMap>}
+     */
+    this._activeActionMaps = new Map();
+
+    /**
+     * actionPath -> { action parameters }
+     * @type {Map<string, Object>}
+     */
+    this._activeActionInfos = new Map();
+
+    // Used during updates
+    this._activations = new Map();
+    this._deactivations = new Map();
+
+    if (loadDefaults) this.loadDefaults();
+  }
+
+  /*
+  @param inputPath {string} a full semantic path for an input, like '/input/keyboard/key/32'
+  @return [inputValue, inputSource]
+  */
+  queryInputPath(inputPath) {
+    for (let inputSourceInfo of this.inputSources) {
+      if (inputPath.startsWith(inputSourceInfo[0]) === false) {
+        continue;
+      }
+      return [inputSourceInfo[1].queryInputPath(inputPath.substring(inputSourceInfo[0].length)), inputSourceInfo[1]];
```

Comment:
Creating arrays per frame is not a good idea, we should find a way to fix it. Maybe as I proposed before, having two functions: `queryInputPathValue` and `queryInputPathInfo` and will return just one value?